runtime/src/rad_claims.rs

```diff
@@ -0,0 +1,494 @@
+            Self::can_update_upload_account(origin)?;
+
+            <UploadAccount<T>>::put(account_id);
+
+            Ok(())
+        }
+
+        /// Stores root hash for correspondent claim merkle tree run
+        ///
+        /// # <weight>
+        /// - Based on origin check and write op
+        /// # </weight>
+        #[weight = 185_000_000]
+        pub fn store_root_hash(origin, root_hash: T::Hash) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            ensure!(Self::get_upload_account() == who, Error::<T>::MustBeAdmin);
+            <RootHashes<T>>::insert(root_hash, true);
+            Self::deposit_event(RawEvent::RootHashStored(root_hash));
+
+            Ok(())
+        }
+
+    }
+}
+
+impl<T: Trait> Module<T> {
+    /// Hash a:b if a < b, else b:a. Uses the runtime module's hasher.
+    pub fn sorted_hash_of(a: &T::Hash, b: &T::Hash) -> T::Hash {
+        let mut h: Vec<u8> = Vec::with_capacity(64);
+        if a < b {
+            h.extend_from_slice(a.as_ref());
+            h.extend_from_slice(b.as_ref());
+        } else {
+            h.extend_from_slice(b.as_ref());
+            h.extend_from_slice(a.as_ref());
+        }
+
+        T::Hashing::hash_of(&h).into()
+    }
+
+    /// Returns true if the given origin can update the upload account
+    fn can_update_upload_account(origin: T::Origin) -> DispatchResult {
+        T::AdminOrigin::try_origin(origin)
+            .map(|_| ())
+            .or_else(ensure_root)?;
+
+        Ok(())
+    }
+}
+
+impl<T: Trait> frame_support::unsigned::ValidateUnsigned for Module<T> {
+    type Call = Call<T>;
+
+    fn validate_unsigned(
+        _source: TransactionSource,
+        call: &Self::Call,
+    ) -> TransactionValidity {
+        if let Call::claim(account_id, amount, sorted_hashes) = call {
+            // Number of proofs should practically never be >30. Checking this
+            // blocks abuse.
+            if sorted_hashes.len() > 30 {
+                return InvalidTransaction::BadProof.into()
+            }
+
+            // Concat account id : amount
+            let mut v: Vec<u8> = account_id.encode();
+            v.extend(amount.encode());
+
+            // Generate root hash
+            let leaf_hash = T::Hashing::hash_of(&v);
+            let mut root_hash = sorted_hashes.iter()
+                .fold(leaf_hash, |acc, hash| Self::sorted_hash_of(&acc, hash));
+
+            // Initial runs might only have trees of single leaves,
+            // in this case leaf_hash is as well root_hash
+            if sorted_hashes.len() == 0 {
+                root_hash = leaf_hash;
+            }
+
+            // Check that root exists in root hash storage
+            if Self::get_root_hash(root_hash) == true {
+                return ValidTransaction::with_tag_prefix("RADclaim")
+                    .priority(T::UnsignedPriority::get())
+                    .longevity(TryInto::<u64>::try_into(
```

Comment:
Does this mean the transaction is valid for `SessionDuration/2` ?

Reply:
Yes, I was thinking that it might be too long, I would like to specify a lifetime that is closer to a signed extrinsic one (which seems that is around 64 blocks)